### ISSUE TITLE
Add `selector-anb-no-unmatchable`

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#5907]
### What is the problem you're trying to solve?

I'd like to catch `nth-*` selectors with zero index like:

```css
a:nth-child(0) {}
```

Although it's valid CSS, it won't match anything.

From [the spec](https://www.w3.org/TR/css-syntax-3/#anb-microsyntax):

> If both _A_ and _B_ are 0, the pseudo-class represents no element in the list.

See: https://twitter.com/dan_abramov/status/1492277543185096704

Related to https://github.com/stylelint/stylelint/issues/5575 (which I suggest we break into separate rules).

### What solution would you like to see?

A new rule to help me [avoid errors](https://stylelint.io/user-guide/rules/list#avoid-errors) in structural pseudo-classes:

- [Tree-Structural pseudo-classes](https://www.w3.org/TR/selectors-4/#structural-pseudos)
- [Grid-Structural Selectors](https://www.w3.org/TR/selectors-4/#table-pseudos)

It'll catch:

```css
a:nth-child(0) {}
```

```css
a:nth-child(+0n) {}
```

```css
a:nth-child(0n+0) {}
```

```css
a:nth-child(0n + 0) {}
```

```css
a:nth-child(0n-0) {}
```

```css
a:nth-child(-0n-0) {}
```

```css
a:nth-of-type(0n) {}
```

It'd also disallow zero index _of S_ selectors, e.g.:

```css
:nth-child(0 of a) {}
```

This is a good candidate for using an _An+B_ parser, e.g. the one from SWC if it's available, otherwise, we can release for just the simple case of `0` as that's the most common mistake.

- Name: `selector-anb-no-unmatchable`
- Primary option: `true`
- Secondary options: None
- Autofixable: No
- Message: `Unexpected unmatchable An+B selector "${puesdoClass}"` (e.g. Unexpected unmatchable An+B selector "nth-child(0)")
- Description: "Disallow unmatchable An+B selectors"
- Extended description: (Something about zero indexes being a common mistake)
- Section: "Avoid errors" -> "Unmatchable"

Any thoughts on the rule, its name and scope?

--- 

Incidentally, there's a set of rules to [enforce (non-stylistic) conventions](https://stylelint.io/user-guide/rules/list#enforce-conventions) in this area too, e.g. preferring:

- `:first-child` over `:nth-child(1)` (or vice versa)
- `:nth-child(even)` over `:nth-child(2n+0)` (or vice versa)
- and so on

If you'd like to add any of these to Stylelint, please open a separate issue.

**Comment by ybiquitous** [comment]
@jeddy3 The new rule sounds good to me. 👍🏼 

I'm concerned about the one thing that the rule name includes a `+` character. We usually use a rule name for a path on file systems and an URL, so such a name may cause any problem.
For example, a `+` character needs to be escaped in an URL.

What do you think?

**Comment by ybiquitous** [comment]
Note: In the spec URL, `anb` is used, instead of `an+b`:

https://drafts.csswg.org/css-syntax/#anb-microsyntax

**Comment by jeddy3** [comment]
`anb` SGTM. I've updated the [blueprint above](https://github.com/stylelint/stylelint/issues/5907#issue-1135539248).

**Comment by jeddy3** [comment]
I've updated the rule name to reflect https://github.com/stylelint/stylelint/issues/5575.

The rule should also catch negative indexes too.

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new rule](https://stylelint.io/developer-guide/rules#add-a-rule) in the Developer guide.

**Comment by mattxwang** [comment]
Hi all, I'm interested in taking this on - have been doing some Rust work on the side, and saw Dan's original tweet.

Were there any follow-up items from #5586 / suggested places to start? I see the Rust code for their `AnPlusB` parser ([in swc-project/swc](https://github.com/swc-project/swc/blob/main/crates/swc_css_parser/src/parser/selectors/mod.rs#L1198)). However, I don't see a public export for it when doing `const swc = require('@swc/core')`. I'm likely missing something!

**Comment by jeddy3** [comment]
> Hi all, I'm interested in taking this on

That's fantastic, thank you. 

We [recently added CSSTree to our dependencies](https://github.com/stylelint/stylelint/pull/6511) for the [`declaration-property-value-no-unknown` rule](https://stylelint.io/user-guide/rules/declaration-property-value-no-unknown/). It [can parse `An+B` syntax](https://astexplorer.net/#/gist/f3b38265d114134243e55db3de18fba0/2e387e7decdb318ca55262700d6709cdf80a2056). We can reach for CSSTree again as this rule doesn't autofix, or we can continue to investigate adding SWC.

It'd be good to see some numbers on the package size, and whether SWC ([like CSSTree](https://github.com/csstree/csstree/blob/master/docs/parsing.md#context)) can parse just selectors.

Regardless of the parser we choose, it's should be pretty trivial to switch, if we ever want to in the future, especially with the tests in place.

(@mattxwang If we decide to go with CSSTree for this rule, please don't feel obliged to take it on as it won't have a Rust aspect which may have piqued your interest.)

**Comment by mattxwang** [comment]
Thanks for the quick response @jeddy3! Didn't realize that CSSTree has this functionality, which is great!

I was able to spin up a first draft for the rule (using CSSTree) in #6678. Let me know what you think! If you'd like, I'm also happy to explore the SWC option, but it seems like CSSTree does the job quite well.
